contracts/stellar-interchain-token-service/src/contract.rs

```diff
@@ -332,6 +332,63 @@ impl InterchainTokenServiceInterface for InterchainTokenService {
         Ok(token_id)
     }
 
+    #[when_not_paused]
+    fn link_token(
+        env: &Env,
+        deployer: Address,
+        salt: BytesN<32>,
+        destination_chain: String,
+        destination_token_address: Bytes,
+        token_manager_type: TokenManagerType,
+        link_params: Option<Bytes>,
+        gas_token: Option<Token>,
+    ) -> Result<BytesN<32>, ContractError> {
+        deployer.require_auth();
+
+        ensure!(
+            !destination_token_address.is_empty(),
+            ContractError::InvalidDestinationAddress
+        );
+
+        // Custom token managers can't be deployed with native interchain token type, which is reserved for interchain tokens
+        ensure!(
+            token_manager_type != TokenManagerType::NativeInterchainToken,
+            ContractError::InvalidTokenManagerType
+        );
+
+        ensure!(
+            destination_chain != Self::chain_name(env),
+            ContractError::InvalidDestinationChain
+        );
```

Comment:
do this generically in `pay_gas_and_call_contract`, if the message type isn't a transfer